### PR TITLE
cleanup AIO quickscript

### PR DESCRIPTION
**Pull request by jagadishmurugan**
- add register provider only if needed
- support passing in CustomLocationOID
- 'az cli' error handling
- set k3s msi to 1.7.639

### CODE REVIEW

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 [hunk 1]
```diff
@@ -12,11 +12,89 @@ param(
...
+    {
+        $connectedCluster = az connectedk8s show $k8sConnectArgs | ConvertFrom-Json
+        if ($connectedCluster.ConnectivityStatus -eq "Connected")
+        {
+            Write-Host "Cluster reached connected status"
+            break
+        }
+
+        Write-Host "Arc connection status is $($connectedCluster.ConnectivityStatus). Waiting for status to be connected..."
+        Start-Sleep -Seconds 10
```
Comment (parameshbabu) [inline tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:45]: is this aggressive polling? probably easier to poll local cluster status with kubectl?

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 [hunk 2]
```diff
@@ -12,11 +12,89 @@ param(
...
+    [object] $arcArgs,
+    [Parameter(Mandatory=$true)]
+    [string] $clusterName
+)
+
+    $k8sConnectArgs = @("-g", $arcArgs.ResourceGroupName)
+    $k8sConnectArgs += @("-n", $clusterName)
+    $k8sConnectArgs += @("--subscription", $arcArgs.SubscriptionId)
+
+    $retries = 150
```
Comment (seanyen) [inline tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:34]: can we define the max retry and interval as a global constant and leave some comments on why we should wait for 25m?

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 [hunk 3]
```diff
@@ -12,11 +12,89 @@ param(
...
+}
+
+function New-ConnectedCluster
+{
+param(
+    [Parameter(Mandatory=$true)]
+    [object] $arcArgs,
+    [Parameter(Mandatory=$true)]
+    [string] $clusterName,
+    [Switch] $useK8s=$false
```
Comment (seanyen) [inline tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:61]: is this switch `useK8s` used anywhere?

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 [hunk 4]
```diff
@@ -182,47 +262,71 @@ if ($retval) {
...
 # Enable custom location support on your cluster using az connectedk8s enable-features command
 Write-Host "Associate Custom location with $ClusterName cluster"
-$objectId = az ad sp show --id bc313c14-388c-4e7d-a58e-70017303ee3b --query id -o tsv
-az connectedk8s enable-features -n $ClusterName -g $ResourceGroupName --custom-locations-oid $objectId --features cluster-connect custom-locations | Out-Null
+$objectId = $aideuserConfigJson.Azure.CustomLocationOID
+if ([string]::IsNullOrEmpty($objectId))
+{
+    $customLocationsAppId = "bc313c14-388c-4e7d-a58e-70017303ee3b"
+    $errOut = $($objectId = & {az ad sp show --id $customLocationsAppId --query id -o tsv}) 2>&1
+    if ($null -eq $objectId)
```
Comment (seanyen) [inline tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:320]: do we want to check `$LASTEXITCODE` as well like how other places invokes azcli?

**Review by seanyen — APPROVED** [review]
LGTM and left some minor comments.

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 (hunk 1 above)
Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:45]: not really, in many cases it shows "Connected" immediately after connect.

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 (hunk 4 above)
Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:320]: in this case returnValue check is better. if it is empty irrespective of LASTEXITCODE we should terminate.

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 (hunk 3 above)
Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:61]: will be used for WorkloadIdentity. passing this in so it is easier to merge later on... WLIF changes are in private branch now...

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 [hunk 5]
```diff
@@ -182,47 +262,71 @@ if ($retval) {
...
 
 # Enable custom location support on your cluster using az connectedk8s enable-features command
 Write-Host "Associate Custom location with $ClusterName cluster"
-$objectId = az ad sp show --id bc313c14-388c-4e7d-a58e-70017303ee3b --query id -o tsv
-az connectedk8s enable-features -n $ClusterName -g $ResourceGroupName --custom-locations-oid $objectId --features cluster-connect custom-locations | Out-Null
+$objectId = $aideuserConfigJson.Azure.CustomLocationOID
+if ([string]::IsNullOrEmpty($objectId))
+{
+    $customLocationsAppId = "bc313c14-388c-4e7d-a58e-70017303ee3b"
+    $errOut = $($objectId = & {az ad sp show --id $customLocationsAppId --query id -o tsv}) 2>&1
```
Comment (parameshbabu) [inline tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:319]: i think this requires owner role to query this oid.

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 (hunk 2 above)
Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:34]: 15m should be sufficient, will add appropriate comments there - thats the recommended timeout from Arc team.

Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1]: prefer to keep the retries local - it is not used elsewhere...

File: tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1 (hunk 5 above)
Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:321]: not just Owner role. It requires Read.All & ReadWrite.All permissions on Graph API...

Comment (jagadishmurugan) [reply tools/scripts/AksEdgeQuickStart/AksEdgeQuickStartForAio.ps1:321]: those are assumed to be done already per AIO team (IT role). now we also allow passing CustomLocationId...

**Review by seanyen — APPROVED** [review]

**Pull request merged**